### PR TITLE
+doc FlexiMerge / FlexiRoute docs

### PR DESCRIPTION
**Pull request by ktoso**
This is a leftover from https://github.com/akka/akka/issues/16549

Will continue to work on this and then `FlexiRoute`.

### CODE REVIEW

**Comment by ktoso** [comment]
Now with completion handling and `FlexiRoute` as well. Found some "I wish I could do X" things, so will open issues in a sec - nothing high priority though - no worries.

**Comment by ktoso** [comment]
// ping @drewhk @patriknw, since you're in docs land anyway now :)

File: akka-docs-dev/rst/scala/stream-customize.rst [hunk 1]
```diff
@@ -149,14 +149,175 @@ Using DetachedStage
 Custom graph processing junctions
 =================================
 
+To extend available fan-in and fan-out structures (graph stages) Akka Streams include :class:`FlexiMerge` and
+:class:`FlexiRoute` which provide an intuitive DSL which allows to describe which upstream or downstream stream
+elements should be pulled from or emitted to.
+
 Using FlexiMerge
 ----------------
+:class:`FlexiMerge` provides an intuitive DSL which can be used to describe from which upstream stage the merge should consume
```
Comment (drewhk) [inline akka-docs-dev/rst/scala/stream-customize.rst]: almost verbatim repetition of the sentence fragment "provides an intuitive DSL which can be used to describe from which upstream" compared to introduction paragraph.

File: akka-docs-dev/rst/scala/stream-customize.rst [hunk 2]
```diff
@@ -149,14 +149,175 @@ Using DetachedStage
...
+Then we change the completion strategy to ``eagerClose`` which will propagate any future completion or error event right
+to this stages downstream effectively shutting down the stream.
+
+.. includecode:: code/docs/stream/FlexiDocSpec.scala#fleximerge-completion
+
+In case you want to change back to the default completion handling, it is available as ``MergeLogic#defaultCompletionHandling``.
 
 Using FlexiRoute
 ----------------
+Similarily to using FlexiMerge, implementing custom fan-out stages requires extending the :class:`FlexiRoute` class
```
Comment (patriknw) [inline akka-docs-dev/rst/scala/stream-customize.rst]: class annotation around FlexiMerge

**Comment by patriknw** [comment]
LGTM

File: akka-docs-dev/rst/scala/stream-customize.rst [hunk 3]
```diff
@@ -149,14 +149,175 @@ Using DetachedStage
...
+available.
+
+Implementing a custom merge stage is done by extending the :class:`FlexiMerge` trait, exposing its input ports and finally
+defining the logic which will decide how this merge should behave. First we need to create the input ports which are used
+to wire up the fan-in element in a :class:`FlowGraph`. These input ports *must* be properly typed and their names should
+indicate what kind of port it is:
+
+.. includecode:: code/docs/stream/FlexiDocSpec.scala#flexi-preferring-merge
+
+Next we implement the ``createMergeLogic`` method, which will be used as factory for the merges :class:`MergeLogic`.
```
Comment (drewhk) [inline akka-docs-dev/rst/scala/stream-customize.rst]: needs rewording: "factory for the merges"

File: akka-docs-dev/rst/scala/stream-customize.rst [hunk 4]
```diff
@@ -149,14 +149,175 @@ Using DetachedStage
...
+A new :class:`MergeLogic` object will be created for each materialized stream, so it is allowed to be stateful.
+
+The :class:`MergeLogic` defines the behaviour of our merge stage, and may be *stateful* (for example to buffer some elements
+internally). The first method we must implement in a merge logic is ``inputHandles`` in which we have the opportunity to
+validate the number of connected input ports, e.g. in our preferring merge we only require that at least one input is connected.
+
+.. warning::
+  While a :class:`MergeLogic` instance *may* be stateful, the :class:`FlexiMerge` instance
```
Comment (drewhk) [inline akka-docs-dev/rst/scala/stream-customize.rst]: good!

File: akka-docs-dev/rst/scala/stream-customize.rst [hunk 5]
```diff
@@ -149,14 +149,175 @@ Using DetachedStage
...
+
+.. includecode:: code/docs/stream/FlexiDocSpec.scala#fleximerge-zip-connecting
+
+.. _flexi-merge-completion-handling-scala:
+
+Completion handling
+^^^^^^^^^^^^^^^^^^^
+Completion handling in :class:`FlexiMerge` is defined by an :class:`CompletionHandling` object which can react on
+completion and error signals from its upstream input ports. The default strategy is to remain running while at-least-one
+upstream input port is still running (i.e. has not signalled completion or error).
```
Comment (drewhk) [inline akka-docs-dev/rst/scala/stream-customize.rst]: well, if you expressed that you only want to read from port X, and port X happens to be closed then it does not remain running?

File: akka-docs-dev/rst/scala/stream-customize.rst [hunk 6]
```diff
@@ -149,14 +149,175 @@ Using DetachedStage
...
+
+Customising completion can be done via overriding the ``MergeLogic#initialCompletionHandling`` method, or from within
+a :class:`State` by calling ``ctx.changeCompletionHandling(handling)``. Other than the default completion handling (as
+late as possible) :class:`FlexiMerge` also provides an ``eagerClose`` completion handling which completes (or errors) its
+downstream as soon as at least one of its upstream inputs completes (or errors).
+
+In the example below the we implement an ``ImportantWithBackups`` fan-in stage which can only keep operating while
+the ``important`` and at-least-one of the ``replica`` inputs are active. Therefore in our custom completion strategy we
```
Comment (drewhk) [inline akka-docs-dev/rst/scala/stream-customize.rst]: nice example

**Comment by drewhk** [comment]
LGTM

File: akka-docs-dev/rst/scala/stream-customize.rst (hunk 2 above)
Comment (ktoso) [inline akka-docs-dev/rst/scala/stream-customize.rst]: thanks

File: akka-docs-dev/rst/scala/stream-customize.rst (hunk 1 above)
Comment (ktoso) [inline akka-docs-dev/rst/scala/stream-customize.rst]: thanks, rewordede

File: akka-docs-dev/rst/scala/stream-customize.rst (hunk 3 above)
Comment (ktoso) [inline akka-docs-dev/rst/scala/stream-customize.rst]: thanks :+1:

File: akka-docs-dev/rst/scala/stream-customize.rst (hunk 5 above)
Comment (ktoso) [inline akka-docs-dev/rst/scala/stream-customize.rst]: thanks! reworded - yes, it completes if the State is unable to be fulfulied

File: akka-docs-dev/rst/scala/stream-customize.rst (hunk 6 above)
Comment (ktoso) [inline akka-docs-dev/rst/scala/stream-customize.rst]: :)

**Pull request merged**